ntoskrnl/mm/ARM3/wslist.cpp

```diff
@@ -0,0 +1,484 @@
+        RemoveFromWsList(WsList, Entry.u1.VirtualAddress);
+
+        /* Dirtify the page, if needed */
+        if (PointerPte->u.Hard.Dirty)
+            Pfn->u3.e1.Modified = 1;
+
+        /* Make this a transition PTE */
+        MI_MAKE_TRANSITION_PTE(PointerPte, Page, Protection);
+        KeInvalidateTlbEntry(MiAddressToPte(PointerPte));
+
+        /* Drop the share count. This will take care of putting it in the standby or modified list. */
+        MiDecrementShareCount(Pfn, Page);
+
+        MiReleasePfnLock(OldIrql);
+        Ret++;
+    }
+    return Ret;
+}
+
+/* GLOBAL FUNCTIONS ***********************************************************/
+extern "C"
+{
+
+VOID
+NTAPI
+MiInsertInWorkingSetList(
+    _Inout_ PMMSUPPORT Vm,
+    _In_ PVOID Address,
+    _In_ ULONG Protection)
+{
+    PMMWSL WsList = Vm->VmWorkingSetList;
+
+    /* Make sure that we are holding the right locks. */
+    ASSERT(MM_ANY_WS_LOCK_HELD(PsGetCurrentThread()));
+    MI_ASSERT_PFN_LOCK_HELD();
+
+    PMMPTE PointerPte = MiAddressToPte(Address);
+
+    /* Make sure we are adding a paged-in address */
+    ASSERT(PointerPte->u.Hard.Valid == 1);
+    PMMPFN Pfn1 = MiGetPfnEntry(PFN_FROM_PTE(PointerPte));
+    ASSERT(Pfn1->u3.e1.PageLocation == ActiveAndValid);
+
+    /* Shared pages not supported yet */
+    ASSERT(Pfn1->u1.WsIndex == 0);
+    ASSERT(Pfn1->u3.e1.PrototypePte == 0);
+
+    /* Nor are "ROS PFN" */
+    ASSERT(MI_IS_ROS_PFN(Pfn1) == FALSE);
+
+    Pfn1->u1.WsIndex = GetFreeWsleIndex(WsList);
+    MMWSLENTRY& NewWsle = WsList->Wsle[Pfn1->u1.WsIndex].u1.e1;
+    NewWsle.VirtualPageNumber = reinterpret_cast<ULONG_PTR>(Address) >> PAGE_SHIFT;
+    NewWsle.Protection = Protection;
+    NewWsle.Direct = 1;
+    NewWsle.Hashed = 0;
+    NewWsle.LockedInMemory = 0;
+    NewWsle.LockedInWs = 0;
+    NewWsle.Age = 0;
+    NewWsle.Valid = 1;
+
+    Vm->WorkingSetSize += PAGE_SIZE;
+    if (Vm->WorkingSetSize > Vm->PeakWorkingSetSize)
+        Vm->PeakWorkingSetSize = Vm->WorkingSetSize;
+}
+
+VOID
+NTAPI
+MiRemoveFromWorkingSetList(
+    _Inout_ PMMSUPPORT Vm,
+    _In_ PVOID Address)
+{
+    RemoveFromWsList(Vm->VmWorkingSetList, Address);
+
+    Vm->WorkingSetSize -= PAGE_SIZE;
+}
+
+_Requires_exclusive_lock_held_(WorkingSet->WorkingSetMutex)
+VOID
+NTAPI
+MiInitializeWorkingSetList(_Inout_ PMMSUPPORT WorkingSet)
+{
+    PMMWSL WsList = WorkingSet->VmWorkingSetList;
+
+    /* Initialize some fields */
+    WsList->FirstFree = ULONG_MAX;
+    WsList->Wsle = reinterpret_cast<PMMWSLE>(WsList + 1);
+    WsList->LastEntry = 0;
+    /* The first page is already allocated */
+    WsList->LastInitializedWsle = (PAGE_SIZE - sizeof(*WsList)) / sizeof(MMWSLE);
+
+    /* Insert the address we already know: our PDE base and the Working Set List */
+    if (MI_IS_PROCESS_WORKING_SET(WorkingSet))
+    {
+        ASSERT(WorkingSet->VmWorkingSetList == MmWorkingSetList);
+#if _MI_PAGING_LEVELS == 4
+        MiInsertInWorkingSetList(WorkingSet, (PVOID)PXE_BASE, 0U);
```

Comment:
The `Protection` parameter is equal to some flag, isn't it?

Reply:
To the Mm internal representation of Nt protection bits. MM_READWRITE & the like. See here:
https://git.reactos.org/?p=reactos.git;a=blob;f=ntoskrnl/mm/ARM3/miarm.h;hb=0dedb9b47457e86ca57b93856af3da6cb0138a93#l35